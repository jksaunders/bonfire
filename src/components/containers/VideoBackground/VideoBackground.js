import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { css } from '../../../styling';
import Layout from '../Layout';

const propTypes = {
  children: PropTypes.node,
  fallbackColor: PropTypes.string,
  filter: PropTypes.string,
  muted: PropTypes.bool,
  parentLayoutProps: PropTypes.shape(Layout.shape).isRequired,
  src: PropTypes.string.isRequired,
};

const defaultProps = {
  children: null,
  fallbackColor: null,
  filter: null,
  muted: true,
};

const Container = styled(Layout)`
  position: relative;
`;

// eslint-disable-next-line react/prop-types
const MutableVideo = ({ className, src, type }) => {
  const ref = useRef();
  const [{ isFirstLoad, videoCanPlay }, setState] = useState({
    isFirstLoad: true,
    videoCanPlay: false,
  });

  useEffect(() => {
    if (ref.current && ref.current.children && ref.current.children[0]) {
      if (ref.current.children[0].play != null) {
        setState({
          isFirstLoad: false,
          videoCanPlay: true,
        });
      } else {
        setState({
          isFirstLoad: false,
        });
      }
    }
  }, [videoCanPlay]);

  const dangerouslySetInnerHTMLProps = {
    dangerouslySetInnerHTML: {
      __html: `
        <video
          autoplay
          loop
          muted
          playsinline
        >
          <source src="${src}" type=${type}>
        </video>
`,
    },
  };

  return (
    <div
      className={className}
      ref={ref}
      // eslint-disable-next-line react/no-danger
      {...(isFirstLoad || videoCanPlay ? dangerouslySetInnerHTMLProps : {})}
      style={{
        ...(!(isFirstLoad || videoCanPlay) ? { padding: '100% 100% 0 0' } : {}),
      }}
    />
  );
};

const Video = styled(MutableVideo)`
  ${css('fallbackColor', 'background-color')}

  > video {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: -100;
    object-fit: cover;
    ${css('filter')}
  }
`;

const VideoBackground = ({
  children,
  fallbackColor,
  filter,
  parentLayoutProps,
  src,
}) => (
  <Container {...parentLayoutProps}>
    <Video
      fallbackColor={fallbackColor}
      filter={filter}
      src={src}
      type="video/mp4"
    />
    {children}
  </Container>
);

VideoBackground.propTypes = propTypes;
VideoBackground.defaultProps = defaultProps;

export default VideoBackground;
