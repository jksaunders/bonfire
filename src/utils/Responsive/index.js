export { default as ResponsiveRoot } from './ResponsiveRoot';
export * from './ResponsiveRoot';
