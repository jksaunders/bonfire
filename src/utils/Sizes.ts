export type SizeProp<T> = Record<string, T>;
