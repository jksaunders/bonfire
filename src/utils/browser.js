const goTo = (url) => {
  if (url && typeof window !== 'undefined') {
    window.location.href = url;
  }
};

export default {
  goTo
};