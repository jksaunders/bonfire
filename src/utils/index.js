export { default as Browser } from "./browser";
export * from "./styling";