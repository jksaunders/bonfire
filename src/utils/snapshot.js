/* eslint-disable import/no-extraneous-dependencies */
import { render as importedRender } from '@testing-library/react';
import 'jest-styled-components';

export const render = component => {
  const rendered = importedRender(component);
  return {
    ...rendered,
    component: rendered.container.firstChild,
  };
};

export const expectSnapshot = component => {
  expect(component).toMatchSnapshot();
};

export const expectExists = component => {
  expect(component).not.toEqual(null);
  expect(component).toBeInTheDocument();
};

export const expectStyle = (component, key, value) => {
  expect(component).toHaveStyleRule(key, value);
};
