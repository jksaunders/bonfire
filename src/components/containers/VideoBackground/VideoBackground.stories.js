import React from "react";
import styled from "styled-components";
import {
  text,
  withKnobs
} from "@storybook/addon-knobs";
import Layout from "../Layout";
import VideoBackground from "./VideoBackground";

export default {
  title: "Containers|Video Background",
  component: VideoBackground,
  decorators: [withKnobs]
};

const Content = styled.div`
  height: ${props => props.height};
  width: ${props => props.width};
`;

export const Basic = () => {
  const heightOfContent = text("Height of content", "200px");
  const widthOfContent = text("Width of content", "200px");
  const filter = text("Filter", "opacity(60%)");

  return (
    <VideoBackground
      filter={filter}
      parentLayoutProps={{ center: true, full: true }}
      src="https://upload.wikimedia.org/wikipedia/commons/6/60/Wikipedia_logo_puzzle_globe_spins_horizontally_and_vertically%2C_revealing_the_contents_of_all_of_its_puzzle_pieces_%284K_resolution%29_%28VP9%29.webm"
    >
      <Content height={heightOfContent} width={widthOfContent}>
        <Layout centered full background="hsla(0,0%,0%,.2)" height="max-content" width="max-content">
          Hi!
        </Layout>
      </Content>
    </VideoBackground>
  );
};