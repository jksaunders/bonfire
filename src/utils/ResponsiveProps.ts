import { SizeProp } from './Sizes';

export type ResponsiveProp<T> = T | SizeProp<T>;
export type ResponsiveProps<T, ResponsiveKeys = keyof T> = {
  [K in keyof T]: K extends ResponsiveKeys
    ? T[K] extends undefined
      ? ResponsiveProp<T[K]>
      : ResponsiveProp<NonNullable<T[K]>>
    : T[K];
};
