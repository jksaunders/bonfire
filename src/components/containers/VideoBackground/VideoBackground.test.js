import React from 'react';
import { expectExists, expectSnapshot, render } from '../../../utils/snapshot';
import VideoBackground from './VideoBackground';

const getVideoBackground = (props = {}) =>
  render(<VideoBackground {...props} />);

test('VideoBackground open', () => {
  const { component } = getVideoBackground({
    filter: 'opacity(70%)',
    muted: false,
    parentLayoutProps: { center: true, full: true },
    src:
      'https://upload.wikimedia.org/wikipedia/commons/6/60/Wikipedia_logo_puzzle_globe_spins_horizontally_and_vertically%2C_revealing_the_contents_of_all_of_its_puzzle_pieces_%284K_resolution%29_%28VP9%29.webm',
  });
  expectExists(component);
  expectSnapshot(component);
});
