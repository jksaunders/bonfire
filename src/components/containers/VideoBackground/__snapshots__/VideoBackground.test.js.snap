// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoBackground open 1`] = `
.c0 {
  box-sizing: border-box;
  height: 100%;
  width: 100%;
}

.c1 {
  display: grid;
  -webkit-font-smoothing: antialiased;
}

.c2 {
  position: relative;
}

.c3 > video {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: -100;
  object-fit: cover;
  -webkit-filter: opacity(70%);
  filter: opacity(70%);
}

<div
  class="c0 c1 c2"
>
  <div
    class="c3"
  >
    
        
    <video
      autoplay=""
      loop=""
      muted=""
      playsinline=""
    >
      
          
      <source
        src="https://upload.wikimedia.org/wikipedia/commons/6/60/Wikipedia_logo_puzzle_globe_spins_horizontally_and_vertically%2C_revealing_the_contents_of_all_of_its_puzzle_pieces_%284K_resolution%29_%28VP9%29.webm"
        type="video/mp4"
      />
      
        
    </video>
    

  </div>
</div>
`;
